Regex search

zwim opened the pull request:
`findText` and friends now search with regular expressions.
Tested with and without soft-hyphens.

For now the regex search is always on -> of course, this will have to be switchable. I am thinking on a prop like `crengine.search.regex`, so we can switch the thing on and off directly from the frontend (without additional changes in base).

Todo: Test more corner cases.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/crengine/442)
<!-- Reviewable:end -->

poire-z commented:
> I am thinking on a prop like crengine.search.regex, so we can switch the thing on and off directly from the frontend (without additional changes in base).

Feels a bit odd: these props are mostly globals. Your regex/non-regex is a function parameter. In base/cre.cpp, we use one of the upper level findText() methods, so it feels best to have this parameter passed via the base function up to the low level findText()s you're tweaking in crengine.

Some other thoughts:
- so, you use C++ standard library std::regex/cmatch ? I know nothing about that :) and crengine is usually avoiding std:: stuff. The only bits of std:: stuff I see are in qtimagescale, and they are std::cerr :)
I don't really know the implications of using them, I let @NiLuJe or @virxkane tell us. (I remember I had once or twice a solution to use some std:: stuff, and had to decide if I would bring this std:: stuff, and decide not to to not be the first to make crengine dependant on stdlib :)
- crengine works with UTF32 - it feels odd (and ugly) to have to convert back and forth the unicode text to utf8 just because these std:: function work in UTF8. Are you sure there are no version/API that works with UTF32 ?
- for counting utf8/utf32 or converting, there are already functions in lvstring.h, and a few private ones in lvstring.cpp (Utf8ByteCount...) - if these would work, you could made them public in lvstring.h, instead of duplicating stuff that do more or less the same things.
- not super sure I understand the flow of your added code - but as it seems at the begining of the existing findText(), isi it possible to put it all in some other functions findTextRegex() that you would call at the beginning of the existing function ? Just for clarity and make it clear they don't interact much? (dunno how this interacts with your `if we have softhyphens in original str, we search for the found pattern later` that I don't really get - you mean it needs some code you haven't yet written ?)

I still have doubts about the usefulness of regexps when we search a book. Forgot to ask in https://github.com/koreader/koreader/issues/7862 for some real-life use cases when this would have felt needed :) Have you ever met a case where this would be needed ? :) (might depend on the language, but in French, I don't see... We have words like "principal/principaux" singular/plural, and one would want to find all variations with `principa..?` :/ (Does the regex code has some pattern to match start/end of word? Should these be added to the pattern by frontend?)

Because it might be fun to implement - but go explain in the frontend to a normal reader what regexp are and the syntax we support.... and it could make our nice app a bit too geeky/techincal.
(I guess the most technical stuff we have are CSS/style tweaks... but it's a bit more justified by its usefulness and the provided style tweaks. This regexp seach doesn't feel so obvious.)

Frenzie commented:
> (I remember I had once or twice a solution to use some std:: stuff, and had to decide if I would bring this std:: stuff, and decide not to to not be the first to make crengine dependant on stdlib :)

That sounds like an odd one to avoid. :-)

I concur that I don't really understand what the use case might be. One can't help but wonder if it might be a bit of an XY problem.

NiLuJe commented:
We link statically to the STL, so the net effect is simply bloat (in terms of code size, and possibly allocations). Then again , it *is* C++, and trying to avoid the STL in 2021 feels like a weird move, but CRe is *old*, and started its life on a platform where compilers and standards were simply wishful thinking ;).

----

And I'm also not seeing the use-case, FWIW ;). I mean, it's "fun", but is it useful?

virxkane commented:
We already have a hard dependency on the "c++0x" standard since https://github.com/buggins/coolreader/pull/183.
Also we in CoolReader use 'c++11' standard for compatibility with Qt5: https://github.com/buggins/coolreader/commit/855c3d18ca599a62c45f25f33bcd6597b5bc5137.
Regex has appeared in С++ since the 'c++11' standard: https://en.cppreference.com/w/cpp/regex, so I don't see anything wrong with using it.

But we are not using the 'stdc++' library in crengine. Nor `std::map`, nor `std::string`, nor `std::wstring` and so on.
When I started contributing to CoolReader when implementing new kerning using HarfBuzz, I used `std::map` for additional cache, but then I saw that there is an `LVHashTable` class in crengine. There is no point in spawning mess and I got rid of using `std::map`: https://github.com/buggins/coolreader/pull/24.
There are many stuff in crengine which are analogous to the stdc++ library. I think one of the options would be to remove all classes from the crengine for which there are analogues in stdc++ and then use stdc++.
But I would not do this for several reasons: 
1. these are not direct analogs, you will have to rewrite a lot of code, for example, when switching from `LVString32` to `std::wstring`;
2. many stuff in crengine are written with an emphasis on high performance, in stdc++ - on maximum portability;
3. I like that some basic things are close by and to see their implementation you don't have to go far, besides, I can always improve them! Of course, this is not possible using a ready-made pre-installed library.

Otherwise I agree with poire-z.

virxkane commented:
> Then again, it _is_ C++, and trying to avoid the STL in 2021 feels like a weird move, but CRe is _old_, and started its life on a platform where compilers and standards were simply wishful thinking ;).

I don't think it's just wishful thinking. It just used the old C++ standard, for example, C++98 or ISO/IEC 14882:1998.

virxkane commented:
Programs using crengine are often used for reading various kinds of literature, so here I don't see why it might need to search using regular expressions. This would be useful, for example, when looking at some kind of configuration files, but this is a rather strange use of CoolReader.
What would really be useful is a search using wordform. For example, you need to find a mention of the russian word "яблоко", i.e. all word forms of this word - "яблока", "яблоке" and so one.
This is a completely different task and seems to require special dictionaries for each language, like hyphenation dictionaries.

zwim commented:
Thanks for all comments.

A use case can be if you want to search "huhn" singular and "hühner" plural, or "Max" "Moritz" "huhn" and "Hühner" at once.

So you can find/highlight words nearby. A dictionary would not help here.

zwim commented:
@poire-z:
> Feels a bit odd: these props are mostly globals. Your regex/non-regex is a function parameter. In base/cre.cpp, we use one of the upper level findText() methods, so it feels best to have this parameter passed via the base function up to the low level findText()s you're tweaking in crengine.

OK, that's fine.

>or counting utf8/utf32 or converting, there are already functions in lvstring.h, and a few private ones in lvstring.cpp (Utf8ByteCount...) - if these would work, you could made them public in lvstring.h, instead of duplicating stuff that do more or less the same things.

I am using as much of the functions available as possible. But I did not find one for counting characters in an utf8 encoded string. -> utf8ByteCount does the opposite of what I need. 

>not super sure I understand the flow of your added code - but as it seems at the begining of the existing findText(), isi it possible to put it all in some other functions findTextRegex() that you would call at the beginning of the existing function ? Just for clarity and make it clear they don't interact much? (dunno how this interacts with your if we have softhyphens in original str, we search for the found pattern later that I don't really get - you mean it needs some code you haven't yet written ?)

I can pull the code out of findText and findTextRev.

>Because it might be fun to implement - but go explain in the frontend to a normal reader what regexp are and the syntax we support.... and it could make our nice app a bit too geeky/techincal.

I would just tell the user that regex can be used (maybe as an own menu entry below "fulltext search"). If someone knows regexps he can use it, all the other can learn it :) Maybe a wiki entry could point to regex?

@Frenzie, @NiLuJe : About code size: crengine is growing about 48k. 

@virxkane 
>    these are not direct analogs, you will have to rewrite a lot of code, for example, when switching from lString32 to std::u32string;
>    many stuff in crengine are written with an emphasis on high performance, in stdc++ - on maximum portability;
>    I like that some basic things are close by and to see their implementation you don't have to go far, besides, I can always improve them! Of course, this is not possible using a ready-made pre-installed library.

I do fully agree.

zwim commented:
I have tested it on German books with umlauts and guillemots and ...

virxkane commented:
> A use case can be if you want to search "huhn" singular and "hühner" plural, or "Max" "Moritz" "huhn" and "Hühner" at once.
> 
> So you can find/highlight words nearby. A dictionary would not help here.

This can be solved by interpreting the search expression as a set of keywords, and substituting all its word forms for each word in turn. In other words, make a primitive search engine. The only problem is to find/compose dictionaries of word forms, at least for the main languages.

Use regular expressions for this...

> "Some people, when confronted with a problem, think I know, I'll use regular expressions. Now they have two problems."

Frenzie commented:
> A use case can be if you want to search "huhn" singular and "hühner" plural, or "Max" "Moritz" "huhn" and "Hühner" at once.

We understand that these are things for which one might use regular expressions. But I'm not sure I quite understand why. You want to quickly assemble some computer-assisted linguistics data on the train and you forgot your laptop? :-)

By the way, what will happen if you enter `.*`? Frozen followed by eventual OOM?

virxkane commented:
@zwim If you still want to force your implementation, try `std::basic_regex<char32_t>` to avoid converting to UTF8. / I don’t know how and whether regex works on non-latin characters. /
Otherwise, try to use the `Utf8CharCount()` function instead of the frankly weak `countChars()`.

zwim commented:
> > A use case can be if you want to search "huhn" singular and "hühner" plural, or "Max" "Moritz" "huhn" and "Hühner" at once.
> 
> We understand that these are things for which one might use regular expressions. But I'm not sure I quite understand why. You want to quickly assemble some computer-assisted linguistics data on the train and you forgot your laptop? :-)
> 
> By the way, what will happen if you enter `.*`? Frozen followed by eventual OOM?

Huh, a good point. This case can be excluded if the found match has a length of zero.

poire-z commented on crengine/include/lvtinydom.h:
```diff
@@ -2014,7 +2014,7 @@ class ldomXRange {
 
     /// searches for specified text inside range
-    bool findText( lString32 pattern, bool caseInsensitive, bool reverse, LVArray<ldomWord> & words, int maxCount, int maxHeight, int maxHeightCheckStartY = -1, bool checkMaxFromStart = false );
+    bool findText( lString32 pattern, bool caseInsensitive, bool reverse, LVArray<ldomWord> & words, int maxCount, int maxHeight, int maxHeightCheckStartY = -1, bool checkMaxFromStart = false, bool useRegex = false );
```
`patternIsRegex` to make it clearer (cf https://github.com/koreader/koreader-base/pull/1378#pullrequestreview-691149223)

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10730,139 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+            tmp += str[i];
+        else
+            soft_hyphens = true;
+    }
+    lString8 str_utf8 = UnicodeToUtf8(tmp); // str[0..pos-1], without softhyphens
+    const char* str_utf8_c_str = str_utf8.c_str();
+
+    std::regex regexp(UnicodeToUtf8(search_pattern).c_str());
+    std::cmatch match;
+
+    int start = 0;
+    int length = 0;
+    bool found = false;
+    while ( countChars(str_utf8_c_str, start) < pos) {
+        if (regex_search(str_utf8_c_str+start, match, regexp)) {
+            found = true;
+            length = match.length(); // in utf8
+            start += match.position()+1; // in utf8
+        } else
+            break;
+    }
+    if (!found)
+        return REGEX_NOT_FOUND;
+
+    --start; // in utf8
+
+    if (!soft_hyphens) {  //no softhyphens, we are ready
+        pos = countChars(str_utf8_c_str, start); //in lString32
+        endpos = pos + countChars(str_utf8_c_str+start, length); // in lString32
+        return REGEX_FOUND;
+    }
+    // if we have softhyphens in original str, we search for the found pattern
+    search_pattern = Utf8ToUnicode(str_utf8_c_str+start, length); // copy match to Unicode
+    return REGEX_FOUND_SOFT_HYPHEN;
 }
 
-static bool findText( const lString32 & str, int & pos, int & endpos, const lString32 & pattern )
+#endif // REGEX_SEARCH
+
+static bool findText( const lString32 & str, int & pos, int & endpos, const lString32 & orig_pattern, bool useRegex )
```
(Not fond of having the bit ugly `orig_pattern` in the function signature but as it's a reference, I guess it would change for the caller too if we update it here - so I guess this is the simplest solution if we want to keep the rest of the function unchanged :)
Or `search_pattern` as you named it above? (less ugly than `orig_pattern` in a function signature, as it sounds like a temp/local variable).

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10730,139 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+static int findRegex( const lString32 & str, int & pos, int & endpos, lString32 & search_pattern )
+{
+    //check if there are soft-hyphs in str
+    bool soft_hyphens = false;
```
`bool has_soft_hyphens`

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10730,139 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+
+/* searches for a regex in str before pos; backwards
+ *
+ * returns
+ *     REGEX_NOT_FOUND if pattern not found
+ *                        no changes to pos, endpos, search_pattern
+ *     REGEX_FOUND     if pattern found and no softhyphens -> ready to go
+ *                        changes pos, endpos
+ *     REGEX_FOUND_SOFT_HYPHEN if pattern found and there are softhyphens
+ *                        changes search_pattern
```
`returns: see above` :)
(to avoid duplicated text and the need to not forget to update it similarly in both places.)

poire-z reviewed (commented):
Well, the regex code is now sufficiently externalize from the original function, so I guess I don't mind the addition :)
Might still be worth investigating `std::basic_regex<char32_t>`.

poire-z commented:
> > By the way, what will happen if you enter .*? Frozen followed by eventual OOM?

I think we don't search across text nodes, but individually in each text node, right?
So, at most, `.*` will match a whole paragraph.

> Huh, a good point. This case can be excluded if the found match has a length of zero.

Will the match really be zero length ?
(Not so fond of having to think about and hack/take care/exclude specific stuff - because it feels like we'll never stop :)

Reading the parameters/options/constants in https://en.cppreference.com/w/cpp/regex/basic_regex, it feels like you'll have to make and hardcode some choices, like the type (basic/extended/grep).
But what about `icase` ? Our full plain text search is case insensitive I think, but if I (rhetorical "I" :) to find `Thomas More` and `Mister More` and `Sir More`, I might want to search for `More` case sensitive and avoid all the numerous `more this` and `more of that` in any english text :)

poire-z commented:
> About code size: crengine is growing about 48k.

145k for me on a kobo/arm build (not tested):
```
3 034 732 libcrengine.so.ORIG
3 179 484 libcrengine.so
```
I guess that if I find these strings in libcrengine.so, it's that it's correctly statically linked?
```
_ZNSt8__detail9_CompilerINSt7__cxx1112regex_traitsIcEEE18_M_expression_termILb0ELb0EEEbRSt4pairIbcERNS_15_BracketMatcherIS3_XT_EXT0_EEE
_ZNSt8__detail9_CompilerINSt7__cxx1112regex_traitsIcEEE25_M_insert_bracket_matcherILb0ELb0EEEvb
_ZTISt23_Sp_counted_ptr_inplaceINSt8__detail4_NFAINSt7__cxx1112regex_traitsIcEEEESaIS5_ELN9__gnu_cxx12_Lock_policyE2EE
_ZTSNSt8__detail11_AnyMatcherINSt7__cxx1112regex_traitsIcEELb0ELb0ELb0EEE_ZNSt11regex_errorD0Ev
_ZNSt11regex_errorC1ENSt15regex_constants10error_typeE
_ZTSSt11regex_error
St23_Sp_counted_ptr_inplaceINSt8__detail4_NFAINSt7__cxx1112regex_traitsIcEEEESaIS5_ELN9__gnu_cxx12_Lock_policyE2EE
NSt8__detail11_AnyMatcherINSt7__cxx1112regex_traitsIcEELb1ELb1ELb1EEE
NSt8__detail12_CharMatcherINSt7__cxx1112regex_traitsIcEELb0ELb0EEE
Number of NFA states exceeds limit. Please use shorter regex string, or use smaller brace expression, or make _GLIBCXX_REGEX_STATE_LIMIT larger.
Unexpected end of regex when in an open parenthesis.
Unexpected end of regex when in brace expression.
Unexpected end of regex when escaping.
Unexpected end of regex when reading control code.
Unexpected end of regex when ascii character.
Unexpected end of regex when in bracket expression.
```

Frenzie commented:
> I think we don't search across text nodes, but individually in each text node, right?

Alright, that's probably safe enough. What if it's a really long text node?

zwim commented:
@poire-z @virxkane : When using std::basic_regex<char32_t> we would need some work (https://en.cppreference.com/w/cpp/regex/regex_traits), as it is not in the standard. I don't want to do that, as I don't see a real benefit.

@Frenzie : Searching on a long .txt document (11MB, one paragraph with 2 million words) is no problem (even with ".*?"). Loading takes longer:) But it seems that crengine splits the long text into paragraphs (don't know why).

@poire-z : '.*?' is non greedy and my return zero length matches, whereas ".*" is greedy and will return the whole string as match.

I would go for hardcoding "Modified ECMAScript regular expression grammar" as this is the default (https://en.cppreference.com/w/cpp/regex/ecmascript).

Yes the search is case insensitive (as the fulltext search in KOReader). I would not change that!

>So, at most, .* will match a whole paragraph.
Yes

>Will the match really be zero length ?
>(Not so fond of having to think about and hack/take care/exclude specific stuff - because it feels like we'll never stop :)

We will stop, I promise. If we get a zero length match, I will skip that in the two functions and if there is no other match I will return "PATTERN_NOT_FOUND".

zwim commented:
>... to find `Thomas More` and `Mister More` and `Sir More`, I might want to search for `More` case sensitive and avoid all the numerous `more this` and `more of that` in any english text :)

In this case I would use the pattern `Thomas More|Mister More|Sir More`.

poire-z commented on crengine/include/lvstring.h:
```diff
@@ -1222,7 +1222,7 @@ void free_ls_storage();
 void CRReinitTimer();
 
-
+int Utf8CharCount( const lChar8 * str, int len );
```
Please put it with the other related ones, on line 1055, below `void Utf8ToUnicode(...)` (with a proper comment like the others :)

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10834,12 +10834,12 @@ static int findRegexRev( const lString32 & str, int & pos, int & endpos, lString
 #endif // REGEX_SEARCH
 
-static bool findText( const lString32 & str, int & pos, int & endpos, const lString32 & orig_pattern, bool useRegex )
+static bool findText( const lString32 & str, int & pos, int & endpos, const lString32 & search_pattern, bool patternIsRegex )
```
Bit ugly this mix of `var_name` and `varName`. The other functions are using camelCase, so do as well: `searchPattern`.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -234,6 +234,12 @@ enum CacheFileBlockType {
 #include <xxhash.h>
 #include <lvtextfm.h>
 
+#define REGEX_SEARCH 1
```
May be, to avoid diverging from upstream (feels like our lvtinydom.cpp should be quite similar) if @virxkane wants to pick this but not enable it, use `#if USE_STD_REGEX == 1` (or a better name for the lib/feature, `USE_STDCPP_REGEX` ?) and add it on our side in koreader-base:
https://github.com/koreader/koreader-base/blob/502bba07e57e652dcf85f0b2a5c6b1d52fbc0f75/thirdparty/kpvcrlib/CMakeLists.txt#L62-L74

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10730,126 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+    lString8 str_utf8 = UnicodeToUtf8(tmp); // str[pos..end], without softhyphens
+    const char* str_utf8_c_str = str_utf8.c_str();
+
+    std::regex regexp(UnicodeToUtf8(searchPattern).c_str(), std::regex::ECMAScript);
+    std::cmatch match;
+
+    if (!regex_search(str_utf8_c_str, match, regexp))
```
What happens if the provided regex is invalid/does not compile? regex_search will just fail, or more bad things happen?
Dunno if that's possible, but if you can catch invalid regex before invoking findText() (that does not return any error code, and I think we should leave it alone :), export a small utility function `bool checkRegex(lString32 pattern, lString32 &error_msg)` so you could possibly use it from frontend to let the user know it's not that there's no match, but it's that its regex is wrong.

Frenzie replied:
I don't think it can be? Or do you mean if some character ended up being invalid utf8 or something.

poire-z replied:
Unbalanced [ or (, standalone * or ?, usually fails:
```python
>>> re.compile("*")
<class 'sre_constants.error'>: nothing to repeat
>>> re.compile("(ab")
<class 'sre_constants.error'>: unbalanced parenthesis
>>> re.compile("[a-")
<class 'sre_constants.error'>: unexpected end of regular expression
```

zwim replied:
Thank you!
`std::regex` constructor throws an exception, we have to catch it.

NiLuJe replied:
Oh, hum. That's... potentially an issue.

Does *anything* in CRe currently throw? Do we even *build* with exceptions enabled?

Mixing C++ exceptions w/ Lua's C errors is... tricky: http://lua-users.org/wiki/ErrorHandlingBetweenLuaAndCplusplus

zwim replied:
It's no  problem to catch the exception in crengine. I have a working version.
I will upload it at 22:00 when I am at home again.

Frenzie replied:
> Do we even build with exceptions enabled?

Now there's an argument against i haven't really considered. That being said, it probably shouldn't be overly problematic in this specific case? Wait and see. ;-D

poire-z replied:
> Does anything in CRe currently throw?

I don't think so.

> Do we even build with exceptions enabled?

No idea.

No idea also if handling exceptions will pull more than the 150Kb I noticed previously, and if a simple` try { regex stuff } catch ( std::runtime_error & e ) { return false; }` does/pulls a lot more than` if (!regex_stuff()) { return false; }`.

virxkane replied:
> Oh, hum. That's... potentially an issue.
> 
> Does _anything_ in CRe currently throw? Do we even _build_ with exceptions enabled?

Using STL and not using exceptions is nonsense.

virxkane replied:
> > Does anything in CRe currently throw?
> 
> I don't think so.

I don't know how critical this is, but
https://github.com/koreader/crengine/blob/fef32929a00406ad2fe73e1ace4ad67c641112f5/crengine/include/lvref.h#L593-L596
https://github.com/koreader/crengine/blob/fef32929a00406ad2fe73e1ace4ad67c641112f5/crengine/include/lvref.h#L615-L618
Most likely, if used correctly, this will never happens :).

poire-z reviewed (commented):
No other comment than these 2 - looks quite simple and non-intrusive.

poire-z commented:
Just in case it feels smaller and safer: https://github.com/kokke/tiny-regex-c (from https://embeddedartistry.com/blog/2019/06/14/regular-expressions-for-embedded-c/).

zwim commented:
The latest commit #156d347 exception handling is enabled. 
For the size issue: on an Android build with regex .`/platform/android/luajit-launcher/assets/libs/libcrengine.so` size is 2573628 without regex it is 2466620 -> So the increase on android will be 104.5kB.

>Just in case it feels smaller and safer: https://github.com/kokke/tiny-regex-c (from https://embeddedartistry.com/blog/2019/06/14/regular-expressions-for-embedded-c/).

I will look into it.

@poire-z: Did you use a debug build?

Not tested case sensitive searches (but the code looks promising, so probably there is no change necessary on crengine).

poire-z commented on crengine/include/lvtinydom.h:
```diff
@@ -45,6 +45,9 @@
 #if MATHML_SUPPORT==1
 #include "mathml.h"
 #endif
+#if USE_STD_REGEX==1
+#include <regex>
+#endif
```
Does this need to be in lvtinydom.h ? (It would then be visible to apps including this, which I don't think it needs to.)

zwim replied:
No. I can move it to lvtinydom.cpp

poire-z commented:
Exact same size as 2 days ago for me on a kobo/arm build (not tested):
```
3 034 732 libcrengine.so.ORIG
3 179 484 libcrengine.so 2 days ago
3 179 484 libcrengine.so today
```

> Did you use a debug build?

I don't think so. I use this in my build script (to build only crengine and what I may want to test on my Kobo, and not the whole koreader):
```sh
export KODEBUG=
DEBUG_NO_DEFAULT=1
MYMAKEFILE=Makefile.cre_arm
cp Makefile $MYMAKEFILE
cat >> $MYMAKEFILE << 'EOT'
mycrengine:
        $(MAKE) -C $(KOR_BASE) $(OUTPUT_DIR)/libs
        $(MAKE) -C $(KOR_BASE) $(LUAJIT)
        $(MAKE) -C $(KOR_BASE) $(OUTPUT_DIR)/libs/libblitbuffer.so
        $(MAKE) -C $(KOR_BASE) $(OUTPUT_DIR)/libs/libkoreader-cre.so

EOT
make TARGET=kobo -f $MYMAKEFILE mycrengine
rm $MYMAKEFILE
```

zwim commented:
> Exact same size as 2 days ago for me on a kobo/arm build (not tested):
> 
> ```
> 3 034 732 libcrengine.so.ORIG
> 3 179 484 libcrengine.so 2 days ago
> 3 179 484 libcrengine.so today
> ```

At least we now, that the exception mechanism does not add additional code :)

poire-z commented:
Dunno, but that's also the case in nightlies: 3.1 MB in the kobo build, only 2.5 MB in the android arm build. No idea why such a difference.

NiLuJe commented:
(Much newer STL versions in the Kobo TC than the hilariously old STL used by GCC in the Android TC).

NiLuJe commented:
> Just in case it feels smaller and safer: https://github.com/kokke/tiny-regex-c (from https://embeddedartistry.com/blog/2019/06/14/regular-expressions-for-embedded-c/).

FWIW, both pcre2 & oniguruma also happen to support UTF-32 natively ;o).

zwim commented:
The last commit improves the backward search with regex a lot.
Minor improvement on forward search on both regular and regex search.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10727,165 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+
+    int left = 0;
+    int right = nb_chars;
+    int start = (left + right)/2;
+    int length = 0;
+    bool found = false;
+    int match_pos = -1;
+    // Doing binary search of the regex from back;
```
No problem if the pattern is right in the middle, i.e. from start-2 to start+2?

zwim replied:
Theoretically no. I will test this case.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10821,15 +10994,15 @@ bool ldomXRange::findText( lString32 pattern, bool caseInsensitive, bool reverse
                 }
                 words.add( ldomWord(_end.getNode(), offs, endpos ) );
-                offs--;
+                offs;
```
No op ?!

zwim replied:
Yes. I will remove it.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10727,176 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+    // a poor mans cache
+    static lString32 last_pattern;
+    static std::regex regexp;
+    if (searchPattern != last_pattern) {
+        last_pattern = searchPattern;
+        if (!generateRegex( searchPattern, regexp))
+            return REGEX_NOT_FOUND;
+    }
+
+    if (!regex_search(str_utf8_c_str, match, regexp))
```
If it fails generateRegex, you'll have it cached, and you'll skip generateRegex() next time, so you'll go on with it as if OK.
(Even if regex_search() handles bad regexp, it feels like some logical issue.)

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10821,15 +11005,14 @@ bool ldomXRange::findText( lString32 pattern, bool caseInsensitive, bool reverse
                         firstFoundTextY = currentTextY;
                 }
                 words.add( ldomWord(_end.getNode(), offs, endpos ) );
-                offs--;
```
Why remove this? What else makes this `while` not infinitely looping ?

zwim replied:
No, the regex search delivers the correct position, in the the traditional findTextRev the position is decreased before start.

poire-z replied:
OK, I see the `pos--` in `findTextRev()`. But that is you that added/moved it there.
Feels a bit less obvious to have it moved there when looking here, you don't easily notice the side effect.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10876,7 +11059,7 @@ bool ldomXRange::findText( lString32 pattern, bool caseInsensitive, bool reverse
                     }
                 }
                 words.add( ldomWord(_start.getNode(), offs, endpos ) );
-                offs++;
+                offs = endpos; // we can go to the next potiential position
```
Dunno how the previous code worked with overlapping matches (looking for "aa" in a "aaaaaaaaaa" string), but this might change this behaviour?

zwim replied:
I did not notice a change with overlapping searches. This seems good

poire-z replied:
Well, there is obviously a change. Searching for `aa` (plain, no regexp):
Before:
![image](https://user-images.githubusercontent.com/24273478/123547264-623e1480-d760-11eb-94f4-93e668f0de71.png)
After:
![image](https://user-images.githubusercontent.com/24273478/123547293-78e46b80-d760-11eb-9358-a4650987e67e.png)

zwim commented:
So far the thing looks good. But one problem consists: There is a problem with finding a single char `.` and utf8. For example If I want to search "Huhn" and Hühn" (notice the umlaut) with a regex like "H.hn" I don't get the umlaut version (Unicode...). 
I will look into this, but it seems that we really really need a 'char32_t' regex.
->My champion would be http://www.akenotsuki.com/misc/srell/en/ which seems to be a drop in replacement?
Setting this PR to draft now.

poire-z commented:
> My champion would be http://www.akenotsuki.com/misc/srell/en/ which seems to be a drop in replacement?

703 KB srell.hpp ?!

zwim commented:
Crengine.so grows about 340kB, but the speed (for inconvenient patterns) increases.
unicode should be no problems now. You can search for [h.hn] and find "huhn", "hühn" ...

Frenzie commented:
Probably best to test some East Asian/Cyllic/Hebrew/Arabic too if reasonably possible. The quickest way might be to generate some from Wikipedia within KOReader, not quite sure. Maybe we should add a few basic samples to our test suite.

poire-z commented:
If we go with srell.hpp, and given it's not dead and [updated quite often](http://www.akenotsuki.com/misc/srell/en/history_en.txt), I think it's best it lies in base like nanosvg (which is also included as a header), so we can bump it independantly from crengine, and we don't pollute crengine git log with manual updates of this file.

poire-z commented on CMakeLists.txt:
```diff
@@ -235,6 +235,7 @@ INCLUDE_DIRECTORIES(
   ${HARFBUZZ_INCLUDE_DIRS}
   ${ANTIWORD_INCLUDE_DIR}
   ${CHM_INCLUDE_DIR}
+  ${SRELL_INCLUDE_DIR}
```
This is not "our" CMakeLists.txt, and not the one we use for building (it's historical, only updates to it are when I picked commits from CoolReader).
Our CMakeLists is the one in base: base/thirdparty/kpvcrlib/CMakeLists.txt

zwim replied:
Thank you, did not know that.

zwim commented on crengine/src/lvtinydom.cpp:
```diff
@@ -238,7 +238,7 @@ enum CacheFileBlockType {
 
 #if USE_STD_REGEX==1
-#include <regex>
+#include <srell.hpp>
```
I don't understand?

NiLuJe replied:
Never mind me, I was reacting to a notification to a single commit that hadn't yet completed the move to base (so I was seeing a duplicate copy of srell.hpp). ;).

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10727,180 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+static bool generateRegex(lString32 & searchPattern, srell::u32regex & regexp)
 {
+    lString32 tmp = removeSoftHyphens(searchPattern);
+    lChar32 *ptr = tmp.modify();
```
```c
value_type * modify() { if (pchunk->nref>1) lock(pchunk->len); return pchunk->buf32; }
const value_type * c_str() const { return pchunk->buf32; }
const value_type * data() const { return pchunk->buf32; }
```
lock() does a bit more work. You're sure you would not be good enough with one of the 2 others?
Or would `srell::u32regex(ptr);` really modify it?

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10876,6 +11063,7 @@ bool ldomXRange::findText( lString32 pattern, bool caseInsensitive, bool reverse
                     }
                 }
                 words.add( ldomWord(_start.getNode(), offs, endpos ) );
+                //offs = endpos; // we can go to the next potiential position
```
Remove this when you're sure you won't go back to it :)

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10727,180 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+        oldPattern = searchPattern;
+    }
+
+    static lString32 old_str;
+    static lString32 str_wo_hyph;
+    static bool has_soft_hyphens;
```
~~static or not static?
I remember @NiLuJe may have tell me a static might be more expensive to access than a local. Dunno how this compares to the little maintenance stuff created for each lString32.~~

Unless they are re-used across calls? Not obvious looking at the code but it seems so. A small comment might help :) Just `// cache these across calls`

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10727,180 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+        old_str = str;
+    }
+
+    lChar32 *str_arr = str_wo_hyph.modify() + pos;
```
Same as above: modify() needed ? (It's usually something for use by the low level lString32 methods.)

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10727,180 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+    bool found = false;
+    int length;
+    int match_pos;
+    // Doing binary search of the regex from back;
+    // Faster than linear search, which will bring doulbe hits.
+    // As a (wanted) sideffect it will reduce matches by minimizes double hits.
```
Just asking: why binary search for backward search, but not for forward search ?
About double hits: no issue with overlapping hits like on my screenshots with normal search ?
Also, I did ask before, but don't remember if you answered: no issue if the pattern is in the middle, spanning over where you start the binary search?

zwim replied:
Forward: The regex searches from str_arr+pos; pos is incremented on every call. So we find always the first match (from str_arr+pos). This is why I wanted `offs=endpos` (this would be faster, but may not find all shifted matches).
Backward: We search the last match in str_arr[0..pos] pos is getting decremented. regex_search delivers the first match. So I start with 'str_arr+mid' and doing binary search for the last position.

If the I have "eeeeeeeeeeeeeeeXXYXXeeeeeeeeeeeeeee" (symmetic)and search for "X.YXXe" the pattern is found.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10767,6 +10768,7 @@ int checkRegex(lString32 & searchPattern)
  *                        changes pos, endpos
  *     REGEX_FOUND_SOFT_HYPHEN if pattern found and there are softhyphens
  *                        changes searchPattern
+ *     REGEX_TO_COMPLEX if the regex does some kind of ReDoS
```
REGEX_ **TOO** _COMPLEX :)
What are the names of the exceptions srell throw in that case? There is an exception always and only when it is too complex?

zwim replied:
According to the docs there are two exceptions thrown from regex_search:
`regex_error(regex_constants::error_complexity)` and `regex_error(regex_constants::error_utf8)`
Shall I catch them separately?

poire-z replied:
No, I guess you may not get error_utf8 when working with utf32 :)

poire-z commented:
There is this in srell.hpp:
```c
#ifdef SRELL_NO_UNICODE_DATA
//  If specified, prevents Unicode data from being output into a
//  resulting binary. Useful when either of icase search or the Unicode
//  property is not needed and want to reduce the size of a binary.
#define SRELL_NO_UNICODE_ICASE
#define SRELL_NO_UNICODE_PROPERTY
#endif
```
As we do the lowercasing before (via lowercase() which uses utf8proc), and "Unicode property" might be only this:
pattern | details
-- | --
\p{...} | Matches any character that has the Unicode property specified in "...". For example, \p{scx=Latin} matches every character defined as a Latin letter in Unicode. This expression can be used also in a character class.
-- | --
\P{...} | Matches any character that does not have the Unicode property specified in "...". This can be used in a character class, too.

Might be worth trying to see how much size we gain.
```diff
 #if USE_SRELL==1
+#define SRELL_NO_UNICODE_DATA
 #include <srell.hpp
 #endif
```

zwim commented:
These are the sizes of crengine.so when playing with srell defines (on android build) in bytes
srell defa                   2745148
no_unicode_icase:     2671420
no_unicode_property: 2593596
no_unicode_data:       2519868
no srell at all:             2466620

First tests with `no_unicode_data` look promising. Obviously I will go for this one :)

poire-z commented on crengine/include/lvtinydom.h:
```diff
@@ -2957,4 +2957,7 @@ void compressCachedData(bool enable);
 // RECT_CACHE_UNPACKED_SPACE and STYLE_CACHE_UNPACKED_SPACE by this factor
 void setStorageMaxUncompressedSizeFactor(float factor);
 
+/// check a regular expression
+int checkRegex(const lString32 & searchPattern);
```
Should be wrapped with `#if USE_SRELL_REGEX == 1` ?
Or have a dummy implementation that returns true or false in lvtinydom.cpp when not USE_SRELL_REGEX .

Also, please check that crengine compiles alright when USE_SRELL_REGEX =0.

zwim replied:
In the next commit I have included a dummy checkRegex function, which simulates the regex is ok. -> Falling back to plaintext search.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10728,224 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+
+    static lString32 oldPattern;
+    static srell::u32regex regexp;
+    // poor mans cache
```
Do here as you did below with the `// poor mans cache of regexp and str_wo_hyphens across calls` comment.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10728,224 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+        int left = 0;
+        int right = pos+1;
+        int start = (left + right)/2;
+        // Doing binary search of the regex from back;
+        // Faster than linear search, which will bring doulbe hits.
+        // As a (wanted) sideffect it will reduce matches by minimizes double hits.
```
(english nitpicks: no ";" needed, "double", "side effect", "by minimiz_ing_ double hits")

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10875,6 +11108,7 @@ bool ldomXRange::findText( lString32 pattern, bool caseInsensitive, bool reverse
                             firstFoundTextY = currentTextY;
                     }
                 }
+
```
(no need to add blank lines in bits of code you ended up not updating :)

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10728,224 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+    int length = 0;
+    int match_pos;
+
+    // only search "^" on beginning of a string
+    if ( searchPattern[0] == '^') {
```
I guess the answer is "no", but just asking to make sure: no need to do anything with "$" and end of string?

zwim replied:
Your are right "no".

poire-z commented:
No more comment.
No afterthought ? No more work needed ?
Let's wait for the frontend PR to be finished before merging and bumping everything, or are you super confident this is done ? :)

zwim commented:
Let us wait until the frontend is finished. We really don't have to hurry.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10741,14 +10741,71 @@ int checkRegex(const lString32 & searchPattern)
+    test = srell::u32regex(U"[*+][^)]*\\)[*+]");
+    try {
+        if (srell::regex_search(regex, match, test))
+            return REGEX_TOO_COMPLEX;
```
Wouldn't you prefer to differentiate those detected by this vs the ones detected by srell ? Using REGEX_TOO_COMPLEX for both seems like we loose some information (about who/why it fails, when we need to understand why).
You could then make the 2 values the same single message in frontend if you don't care there.

zwim replied:
It is no problem to add another error constant. 
But if we do not detect the evil regex before search, the search would give us the same error code (after some time). TOO_COMPLEX. So this function is just a shortcut to the error.

> (Feels a bit like over-engineering, and more occasions for the need to update/fix this code.)

If you enter the regex `[\W*|[^ ]*s[^ ]* the emulator searches for about a minute! Don't want to benchmark this on my Tolino :)
Even with the hourglass every user would think the program is frozen.

poire-z replied:
> so we can feed other potential disastrous regex to the checkRegex function directly from the frontend. I will sleep about this idea.

Well, if you really want to check for evil regexs, it's best its kept part of crengine, so it's self sufficient. Feedling that from frontend just to avoid some crengine bumps is not the best idea - it just splits the mess/stuff in 2 places :)

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10741,14 +10741,71 @@ int checkRegex(const lString32 & searchPattern)
+    }
+
+    // detect regex like '(aa|a)', '(aaa|a?)' or '(\w|[^ ])'
+    test = srell::u32regex(U".*(?<left>[^\\(][^|]+)\\|(?<right>[^\\|][^\\)]+).*");
```
Can't this and (the other ones) be made static ? To avoid having to compile them at each lookup ?

zwim replied:
Yes, good point.

poire-z reviewed (commented):
(Feels a bit like over-engineering, and more occasions for the need to update/fix this code.)

zwim commented:
In the last commit I dropped detections of `(a|aa)` patterns as these would give false positives on `(\w|[^ ])` which I did not detect before I fixed lowercase search not to lowercase `\W` and friends.

I lowerered `regexp.limit_counter` to 1<<15, which would do no change on simple patterns like `Sir More|Michael More|Mister More|Mr\. More`. On longer running regexes there will be the error message that the pattern is too complex and not all hits are shown.

On evil expressions the maximum search time will be acceptable, so the user does not think KOReader has frozen.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10750,6 +10755,16 @@ int clearRegexSearchError(int clear)
+            ++i; //skip next character e.g. `\W`, `\w`
+            continue;
+        }
+        pattern[i] = tolower(pattern[i]);
```
To avoid duplicating lStr_lowercase() 's code as tolower(), may be here something like this can work:
```c
lChar32 * pattern_str = pattern.modify();
for (...)
    lStr_lowercase(pattern+i, 1);
```
Or something like this.

zwim replied:
Works.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10742,6 +10742,11 @@ int clearRegexSearchError(int clear)
     return 0; // no error
 }
 
+void lowercasePattern(lString32 & pattern)
+{
+    return;
+}
```
Ugly :)
This looks like an internal function, so better wrap when it is used in a `#if USE_SRELL_REGEX == 1`.
And even inline the non-empty one where it is used, as it seems only used once.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -11128,8 +11142,12 @@ static bool findTextRev( const lString32 & str, int & pos, int & endpos, const l
-    if ( caseInsensitive )
-        pattern.lowercase();
+    if ( caseInsensitive ) {
+        if ( !patternIsRegex )
+            pattern.lowercase();
+        else
+            lowercasePattern( pattern );
```
Better put everything here wrapped in `#if USE_SRELL_REGEX == 1` (even if it adds a bit of noise here).

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10737,11 +10737,29 @@ int checkRegex(const lString32 & searchPattern)
+static int regexSearchError = 0;
+
+// get and clear regexSearchError
+int clearRegexSearchError()
```
This name all around up to base is so weird. `getAndClearLastRegexSearchError()` feels better.
Does it have to be a getAndClear...() ? Can't it be just get...() , and you would clear it when doing a new search ? Or does frontend really needs it to be get & forget?

zwim replied:
It has to be a get and clear functionality, because the findRegex function is called often and not always returns early due to a too complex regex.

The frontend needs this to know so that it can display a notification about missed hits.

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10756,36 +10774,35 @@ int checkRegex(const lString32 & searchPattern)
+            return true;
     } catch (...) {
-        assert( 1 == 2);
+        return false; // allthough never seen, this could be caused by an SRELL internal error.
```
(`although` with a single L - happens to me often to write it with double L and ending T...)

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10728,309 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+    return range.findText( pattern, caseInsensitive, reverse, words, maxCount, maxHeight, maxHeightCheckStartY, false, patternIsRegex );
+}
+
+#if USE_SRELL_REGEX !=1
+int checkRegex(const lString32 & searchPattern)
+{
+    return 0; // no error
+}
+
+int getAndClearRegexSearchError()
+{
+    return 0; // no error
+}
```
I think I'd prefer to not have these dummy implementations, and have the prototypes in lvtinydom.h themselves wrapped in a `#if USE_SRELL_REGEX == 1`.
(And if we were to have these dummy implementations, they should probably return an error instead of pretending all is fine :)

zwim replied:
If we wrap the functions in `lvtinydom.h` we have the problem, that when compiling `cre.cpp` the `-DUSE_SRELL_REGEX` is not defined.

poire-z replied:
So may be with this ?
```diff
--- a/Makefile
+++ b/Makefile
@@ -154,9 +154,9 @@ endif
 $(OUTPUT_DIR)/libs/libkoreader-cre.so: cre.cpp \
                        $(if $(USE_LUAJIT_LIB),$(LUAJIT_LIB),) \
                        $(CRENGINE_LIB)
        $(CXX) -I$(CRENGINE_SRC_DIR)/crengine/include/ $(DYNLIB_CXXFLAGS) \
-               -DLDOM_USE_OWN_MEM_MAN=$(if $(WIN32),0,1) \
+               -DLDOM_USE_OWN_MEM_MAN=$(if $(WIN32),0,1) -DUSE_SRELL_REGEX=1 \
                $(if $(WIN32),-DQT_GL=1) $(SYMVIS_FLAGS) -static-libstdc++ -o $@ $^
 ifdef DARWIN
        install_name_tool -change \
                `otool -L "$@" | grep "libluajit" | awk '{print $$1}'` \
```
It adds some noise here, but it feels better than having it that strange in crengine.

zwim replied:
Shall I add a `-DUSE_SRELL_REGEX` to the base Makefile Line 159
```
$(OUTPUT_DIR)/libs/libkoreader-cre.so: cre.cpp \
                        $(if $(USE_LUAJIT_LIB),$(LUAJIT_LIB),) \
                        $(CRENGINE_LIB)
        $(CXX) -I$(CRENGINE_SRC_DIR)/crengine/include/ $(DYNLIB_CXXFLAGS) \
                -DLDOM_USE_OWN_MEM_MAN=$(if $(WIN32),0,1) \
                -DUSE_SRELL_REGEX \
                $(if $(WIN32),-DQT_GL=1) $(SYMVIS_FLAGS) -static-libstdc++ -o $@ $^
ifdef DARWIN
```

poire-z replied:
Yep (see my previous comment :). But with `=1` as you check for `-DUSE_SRELL_REGEX=1`

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10724,11 +10728,309 @@ bool ldomDocument::findText( lString32 pattern, bool caseInsensitive, bool rever
+            regexSearchError = REGEX_TOO_COMPLEX;
+            return false;
+        }
+
```
(Uneeded blank line.)

poire-z commented on crengine/src/lvtinydom.cpp:
```diff
@@ -10757,8 +11059,26 @@ static bool findText( const lString32 & str, int & pos, int & endpos, const lStr
+            regexSearchError = REGEX_TOO_COMPLEX;
+            return false;
+        }
+
```
(Uneeded blank line.)

poire-z commented on crengine/include/lvtinydom.h:
```diff
@@ -2957,9 +2957,14 @@ void compressCachedData(bool enable);
 int checkRegex(const lString32 & searchPattern);
 /// clear the last Regex search result; return last error
 int getAndClearRegexSearchError();
+#else
+inline int checkRegex(const lString32 & searchPattern) { return -1; }
+inline int getAndClearRegexSearchError() { return -1; }
+#endif
```
Why do we need these ?
Looks to me we don't need them (and frontend users don't have to have them if they don't define USE_SRELL_REGEX).

zwim replied:
We don't need them, if USE_SRELL_REGEX is not set. But then we would also #ifdef the cre.cpp functions `checkRegex` and `getAndClearRegexSearchError`.

zwim commented:
Good news. Today is not good, as I have to leave soon. But tomorrow afternoon-evening is fine.

poire-z commented:
(Ping me when we can go on.)

zwim commented:
@poire-z: I am ready.
